### PR TITLE
Add matrix inverse

### PR DESCRIPTION
**Pull request by lzhao4ever**
for CpuMatrix using clapack
and GpuMatrix using culapack.

### CODE REVIEW

File: paddle/cuda/src/hl_cuda_cublas.cc [hunk 1]
```diff
@@ -162,6 +169,65 @@ void hl_matrix_transpose(real *A_d, real *C_d, int dimM, int dimN) {
...
+               info_d, nMatrices));
+
+  hl_memcpy(info_h, info_d, nMatrices*sizeof(int));
+  for (size_t i = 0; i < nMatrices; i++) {
+    if (info_h[i] != 0) {
+      LOG(FATAL) << "Factorization of matrix failed: matrix may be singular.\n";
+    }
+  }
+
+  /* Step 2: Computer the inverse of the matrix given its LU decomposition */
```
Comment (emailweixu) [inline paddle/cuda/src/hl_cuda_cublas.cc]: computer => compute

File: paddle/cuda/src/hl_cuda_cublas.cc [hunk 2]
```diff
@@ -162,6 +169,65 @@ void hl_matrix_transpose(real *A_d, real *C_d, int dimM, int dimN) {
...
+  real **out_h = (real **)hl_malloc_host(nMatrices*sizeof(real *));
+  for (size_t i = 0; i < nMatrices; i++) {
+    out_h[i] = (real *)((char*)C_d+i*((size_t)dimN*ldc)*sizeof(real));
```
Comment (emailweixu) [inline paddle/cuda/src/hl_cuda_cublas.cc]: out_h[i] = C_d + i * dimN * ldc;

File: paddle/cuda/include/hl_cuda_cublas.h [hunk 3]
```diff
@@ -50,6 +50,22 @@ extern void hl_matrix_transpose(real *A_d,
                                 int dimM,
                                 int dimN);
 
+/*
+ * @brief Matrix inverse
+ *
+ * @param[in]   A_d    input matrix (M x M).
```
Comment (emailweixu) [inline paddle/cuda/include/hl_cuda_cublas.h]: M x M => dimN * dimN

File: paddle/math/Matrix.h [hunk 4]
```diff
@@ -328,6 +328,11 @@ class Matrix : public BaseMatrix {
     LOG(FATAL) << "Not implemented";
   }
 
+  virtual MatrixPtr getInverse() = 0;
+  virtual void inverse(MatrixPtr matInv, bool memAlloc) {
```
Comment (emailweixu) [inline paddle/math/Matrix.h]: comment memAlloc

File: paddle/math/MathFunctions.cpp [hunk 5]
```diff
@@ -40,6 +40,46 @@ void gemm<double>(const CBLAS_TRANSPOSE transA, const CBLAS_TRANSPOSE transB,
 }
 
 template<>
+int getrf<float>(const CBLAS_ORDER Order, const int M, const int N,
```
Comment (hedaoyuan) [inline paddle/math/MathFunctions.cpp]: Order not used.

File: paddle/math/Matrix.cpp [hunk 6]
```diff
@@ -1420,6 +1444,48 @@ void CpuMatrix::transpose(MatrixPtr matTrans, bool memAlloc) {
...
+    CHECK_NE(*data, 0);
+    *dataInv = 1.0 / (*data);
+    return;
+  }
+
+  /* Compute the LU decomposition of the matrix */
+  int *ipiv = new int[height_];
+  CHECK(ipiv != NULL);
+  CBLAS_ORDER order = (matInv->isTransposed() ? CblasColMajor : CblasRowMajor);
+  int info = getrf<real>(order, height_, height_, dataInv, ldc, ipiv);
```
Comment (hedaoyuan) [inline paddle/math/Matrix.cpp]: order is not used in the getrf api.

File: paddle/math/SparseMatrix.h [hunk 7]
```diff
@@ -91,6 +91,11 @@ class GpuSparseMatrix : public Matrix {
   /// B = A'
   void transpose(MatrixPtr matTrans, bool memAlloc);
 
+  MatrixPtr getInverse() { LOG(FATAL) << "Not implemented"; return NULL; }
```
Comment (hedaoyuan) [inline paddle/math/SparseMatrix.h]: if not implemented, do not need define here.

File: paddle/math/CpuSparseMatrix.h [hunk 8]
```diff
@@ -43,6 +43,11 @@ class CpuSparseMatrix : public Matrix {
 
   MatrixPtr getTranspose();
 
+  MatrixPtr getInverse() { LOG(FATAL) << "Not implemented"; return NULL; }
```
Comment (hedaoyuan) [inline paddle/math/CpuSparseMatrix.h]: if not implemented, do not need define here.

File: paddle/cuda/src/hl_cuda_cublas.cc [hunk 9]
```diff
@@ -162,6 +169,65 @@ void hl_matrix_transpose(real *A_d, real *C_d, int dimM, int dimN) {
   hl_matrix_transpose(A_d, C_d, dimM, dimN, dimN, dimM);
 }
 
+void hl_matrix_inverse(real *A_d, real *C_d, int dimN, int lda, int ldc) {
+  /* Solve Ax = I */
+  CHECK_NOTNULL(A_d);
+  CHECK_NOTNULL(C_d);
+
+  /* Step 1: Compute the LU decomposition of matrix A */
+  const unsigned int nMatrices = 1;
```
Comment (hedaoyuan) [inline paddle/cuda/src/hl_cuda_cublas.cc]: If this api consider only calculating one matrix. Remove `nMatrices` can make the code more clear.

File: paddle/cuda/src/hl_cuda_cublas.cc [hunk 10]
```diff
@@ -162,6 +169,65 @@ void hl_matrix_transpose(real *A_d, real *C_d, int dimM, int dimN) {
...
+
+  real **inout_h = (real **)hl_malloc_host(nMatrices*sizeof(real *));
```
Comment (hedaoyuan) [inline paddle/cuda/src/hl_cuda_cublas.cc]: If you want to use a small temporary memory. 
`t_resource.cpu_mem` and `t_resource.gpu_mem` can be used (size is `HPPL_GPU_MEMORY_SIZE`).

File: paddle/cuda/src/hl_cuda_cublas.cc [hunk 11]
```diff
@@ -162,6 +169,65 @@ void hl_matrix_transpose(real *A_d, real *C_d, int dimM, int dimN) {
...
+
+  int *pivot_d = (int *)hl_malloc_device(dimN*nMatrices*sizeof(int));  
+  int *info_h = (int *)hl_malloc_host(nMatrices*sizeof(int));
+  int *info_d = (int *)hl_malloc_device(nMatrices*sizeof(int));
+
+  CHECK_CUBLAS(CUBLAS_GETRF(t_resource.handle,
+	       dimN, inout_d, lda, pivot_d,
+               info_d, nMatrices));
+
+  hl_memcpy(info_h, info_d, nMatrices*sizeof(int));
```
Comment (hedaoyuan) [inline paddle/cuda/src/hl_cuda_cublas.cc]: Can be checked in gpu code. Do not need copy to cpu.

File: paddle/cuda/src/hl_cuda_cublas.cc [hunk 12]
```diff
@@ -162,6 +169,65 @@ void hl_matrix_transpose(real *A_d, real *C_d, int dimM, int dimN) {
   hl_matrix_transpose(A_d, C_d, dimM, dimN, dimN, dimM);
 }
 
+void hl_matrix_inverse(real *A_d, real *C_d, int dimN, int lda, int ldc) {
```
Comment (hedaoyuan) [inline paddle/cuda/src/hl_cuda_cublas.cc]: cublasSgetrfBatched is used to calculate a number of small-sized matrices. 
So, it is best to comment that the performance of this interface may be poor.

Comment (lzhao4ever) [inline paddle/cuda/src/hl_cuda_cublas.cc]: Do you have recommendation of an interface with better performance?

File: paddle/math/CpuSparseMatrix.h (hunk 8 above)
Comment (lzhao4ever) [inline paddle/math/CpuSparseMatrix.h]: It's defined as a pure virtual function in its base class. It I do not define it here, I cannot allocate an object of abstract type.

File: paddle/math/SparseMatrix.h (hunk 7 above)
Comment (lzhao4ever) [inline paddle/math/SparseMatrix.h]: See below.

File: paddle/cuda/src/hl_cuda_cublas.cc (hunk 11 above)
Comment (lzhao4ever) [inline paddle/cuda/src/hl_cuda_cublas.cc]: How to check in gpu data

File: paddle/math/CpuSparseMatrix.h (hunk 8 above)
Comment (hedaoyuan) [inline paddle/math/CpuSparseMatrix.h]: `CpuSparseMatrix` and `GpuSparseMatrix` are no need to define functions that are not implemented. You can see many similar implementations in `Matrix`(like softmax, sequenceSoftmax.... They are not pure virtual functions, the same as the `getInverse` which you add in the `Matrix`).

File: paddle/cuda/src/hl_cuda_cublas.cc (hunk 10 above)
Comment (hedaoyuan) [inline paddle/cuda/src/hl_cuda_cublas.cc]: `hl_vector_sum` is an example.

File: paddle/cuda/src/hl_cuda_cublas.cc (hunk 12 above)
Comment (lzhao4ever) [inline paddle/cuda/src/hl_cuda_cublas.cc]: Done.

File: paddle/math/Matrix.cpp [hunk 13]
```diff
@@ -1420,6 +1444,48 @@ void CpuMatrix::transpose(MatrixPtr matTrans, bool memAlloc) {
...
+void CpuMatrix::inverse(MatrixPtr matInv, bool memAlloc) {
+  CHECK_EQ(height_, width_);
+
+  if (memAlloc) {
+    matInv = std::make_shared<CpuMatrix>(height_, width_);
+  } else {
+    CHECK(matInv != NULL);
+  }
+
+  CHECK_EQ(height_, matInv->getHeight());
```
Comment (lzhao4ever) [inline paddle/math/Matrix.cpp]: Done.

**Pull request merged**